### DISCUSSION TITLE
Where to ref.watch in Notifier

### DISCUSSION THREAD

**Discussion by markbeij** [rrousselGit/riverpod#3347] [Q&A] [answered]
I'm trying to understand what the correct way is to watch a provider from within a notifier of a different provider.

Does this..
```dart
class SomeNotifier extends AsyncNotifier<List<Item>?> {
  ItemsControllerApi get api =>
      ref.watch(apiClientProvider).getItemControllerApi();

  @override
  FutureOr<List<Item>?> build() {
    return null;
  }

  Future<void> load() async {
    state = const AsyncLoading();
    final result = await api.getValueUsingGET();
    state = AsyncData(result.data);
  }
}
```
..do exactly the same as this?
```dart
class SomeNotifier extends AsyncNotifier<List<Item>?> {
  late final ItemsControllerApi api;

  @override
  FutureOr<List<Item>?> build() {
    api = ref.watch(apiClientProvider).getItemControllerApi();
    return null;
  }

  Future<void> load() async {
    state = const AsyncLoading();
    final result = await api.getValueUsingGET();
    state = AsyncData(result.data);
  }
}
```

I like the first variant because it looks cleaner, but 
* Is the effect the same?
* Is it supported?
* Is it recommended?

I can't find this in the documentation.. thanks for anyone sharing their thoughts

**Comment by rrousselGit** [comment, marked as answer]
Only use `ref.watch` inside the `build` method. That's what's said by the documentation :)

**Reply by markbeij** [reply]
I think you're referring to the following sentence in the [documentation](https://riverpod.dev/docs/concepts/reading#using-refwatch-to-observe-a-provider):
> ref.watch is used inside the build method of a widget or inside the body of a provider to have the widget/provider listen to a provider

Thanks for making this explicit.

**Comment by Trung15010802** [comment]
So we should use the second right? @markbeij

**Reply by markbeij** [reply]
Yes

**Comment by gdurandrexel** [comment]
You should use watch (inside a build method) when you want to rebuild when the watched provider has its state updated.

The 2nd example you provide is quite strange. Your build method always returns null whatever the state of the provider you watch, so basically it resets the state of SomeNotifier to null each time apiClientProvider's state changes. Maybe it's what you want, but it's unusual.
If you just want to delay the loading until you call `load`, you can avoid watching apiClientProvider, and just call `final result = await ref.read(apiClientProvider).getItemControllerApi().getValueUsingGET();` in `load()`, so basically your first example, but with `read` instead of `watch`.
